src/core/ModuleManager.sol

```diff
@@ -165,58 +172,105 @@ abstract contract ModuleManager is AccountBase, Receiver {
         internal
         virtual
     {
-        address currentFallback = _getModuleManagerStorage().fallbackHandler;
-        if (currentFallback != fallbackHandler) revert InvalidModule(fallbackHandler);
-        _getModuleManagerStorage().fallbackHandler = address(0);
-        IFallback(currentFallback).onUninstall(initData);
+        // address currentFallback = _getModuleManagerStorage().fallbackHandler;
+        // if (currentFallback != fallbackHandler) revert InvalidModule(fallbackHandler);
+        // _getModuleManagerStorage().fallbackHandler = address(0);
+        // IFallback(currentFallback).onUninstall(initData);
     }
 
-    function _isFallbackHandlerInstalled() internal view virtual returns (bool) {
-        return _getModuleManagerStorage().fallbackHandler != address(0);
+    function _isFallbackHandlerInstalled(bytes4 functionSig) internal view virtual returns (bool) {
+        FallbackHandler storage $fallback = $moduleManager().$fallbacks[functionSig];
+        return $fallback.handler != address(0);
     }
 
-    function _isFallbackHandlerInstalled(address _handler) internal view virtual returns (bool) {
-        return _getModuleManagerStorage().fallbackHandler == _handler;
+    function _isFallbackHandlerInstalled(
+        bytes4 functionSig,
+        address _handler
+    )
+        internal
+        view
+        virtual
+        returns (bool)
+    {
+        FallbackHandler storage $fallback = $moduleManager().$fallbacks[functionSig];
+        return $fallback.handler == _handler;
     }
 
-    function getActiveFallbackHandler() external view virtual returns (address) {
-        return _getModuleManagerStorage().fallbackHandler;
+    function getActiveFallbackHandler(bytes4 functionSig)
+        external
+        view
+        virtual
+        returns (FallbackHandler memory)
+    {
+        return $moduleManager().$fallbacks[functionSig];
     }
 
     // FALLBACK
     fallback() external payable override(Receiver) receiverFallback {
-        address handler = _getModuleManagerStorage().fallbackHandler;
-        if (handler == address(0)) revert();
-        /* solhint-disable no-inline-assembly */
-        /// @solidity memory-safe-assembly
-        assembly {
-            // When compiled with the optimizer, the compiler relies on a certain assumptions on how
-            // the
-            // memory is used, therefore we need to guarantee memory safety (keeping the free memory
-            // point 0x40 slot intact,
-            // not going beyond the scratch space, etc)
-            // Solidity docs: https://docs.soliditylang.org/en/latest/assembly.html#memory-safety
-            function allocate(length) -> pos {
-                pos := mload(0x40)
-                mstore(0x40, add(pos, length))
+        FallbackHandler storage $fallbackHandler = $moduleManager().$fallbacks[msg.sig];
+        address handler = $fallbackHandler.handler;
+        CallType calltype = $fallbackHandler.calltype;
```

Comment:
Just out of curiosity, what does the Solidity compiler do here? Does it know to optimize this into a single SLOAD?

Reply:
If I remember correctly it does. But worth double checking. or writing in `assembly`.

Will do and let you know :)

Reply:
didnt investigate properly, but the gas diff when hardcoding the handler is `92 gas` so it seems like there is just one SLOAD